identity-iota/src/credential/credential_validator.rs

```diff
@@ -0,0 +1,645 @@
+use identity_credential::credential::Credential;
+use identity_did::verifiable::VerifierOptions;
+use serde::Serialize;
+
+use super::errors::ValidationError;
+use super::CredentialValidationOptions;
+
+/// A struct for validating [Credential]s.
+#[non_exhaustive]
+pub struct CredentialValidator;
+
+// Use the following pattern throughout: for each public method there is corresponding private method returning a more
+// local error.
+type ValidationUnitResult = std::result::Result<(), ValidationError>;
+type CredentialValidationResult = std::result::Result<(), AccumulatedCredentialValidationError>;
+
+impl Default for CredentialValidator {
+  fn default() -> Self {
+    Self::new()
+  }
+}
+impl CredentialValidator {
+  /// Constructs a new [CredentialValidator]
+  pub fn new() -> Self {
+    Self {}
+  }
+
+  /// Validates the semantic structure of the [Credential].
+  ///
+  /// # Terminology
+  /// This is a *validation unit*
+  pub fn check_structure<T>(credential: &Credential<T>) -> Result<()> {
+    Self::check_structure_local_error(credential).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn check_structure_local_error<T>(credential: &Credential<T>) -> ValidationUnitResult {
+    credential
+      .check_structure()
+      .map_err(ValidationError::CredentialStructure)
+  }
+
+  /// Validate that the [Credential] does not expire before the specified [Timestamp].
+  ///
+  /// # Terminology
+  /// This is a *validation unit*
+  pub fn does_not_expire_before<T>(credential: &Credential<T>, timestamp: Timestamp) -> Result<()> {
+    Self::does_not_expire_before_local_error(credential, timestamp).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn does_not_expire_before_local_error<T>(credential: &Credential<T>, timestamp: Timestamp) -> ValidationUnitResult {
+    let is_ok = if let Some(expiration_date) = credential.expiration_date {
+      expiration_date >= timestamp
+    } else {
+      true
+    };
+    is_ok.then(|| ()).ok_or(ValidationError::ExpirationDate)
+  }
+
+  /// Validate that the [Credential] is issued no later than the specified [Timestamp].
+  ///
+  /// # Terminology
+  /// This is a *validation unit*
+  pub fn was_not_issued_after<T>(credential: &Credential<T>, timestamp: Timestamp) -> Result<()> {
+    Self::was_not_issued_after_local_error(credential, timestamp).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn was_not_issued_after_local_error<T>(credential: &Credential<T>, timestamp: Timestamp) -> ValidationUnitResult {
+    (credential.issuance_date <= timestamp)
+      .then(|| ())
+      .ok_or(ValidationError::IssuanceDate)
+  }
+
+  // helper method for verifying the signature of a trusted issuer
+  fn extract_corresponding_issuer_helper<'a, T, I: Iterator<Item = &'a ResolvedIotaDocument>>(
+    credential: &Credential<T>,
+    mut issuer_docs: I,
+  ) -> std::result::Result<&'a ResolvedIotaDocument, ValidationError> {
+    let issuer_did: Result<IotaDID> = credential.issuer.url().as_str().parse();
```

Comment:
I think this function can be inlined. It's only used in one location and the iterator seems a bit overkill now?